File: 81. Search in Rotated Sorted Array II/Solution.java
class Solution {
    public int search(int[] nums, int target) {
        int pivot = findPivot(nums);
        // no pivot present that means array not rotated normal binary search 
        if(pivot == -1){
            return binarySearch(nums, 0, nums.length-1, target);
        }
        // if pivot found we have two ascending arrays
        if(nums[pivot] == target){
            return pivot;
        } else if (target>=nums[0]){
            return binarySearch(nums, 0, pivot-1, target);
        } else {
            return binarySearch(nums, pivot+1, nums.length-1, target);
        }
    }

    int findPivot(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        // array not rotated
        if (nums[start] <= nums[end]) {
            return -1;
        }
        while(start <= end){
            int mid = start + (end - start) / 2;
            if(mid < end && nums[mid] > nums[mid+1]){
                return mid;
            }
            if(mid > start && nums[mid] < nums[mid-1]){
                return mid-1;
            } 
            // if elements at mid start and end are equal then skip duplicates
            if(nums[mid] == nums[start] && nums[mid] == nums[end]){
                // skip the start and end 
                // but what if start and end are the pivots 
                // check for pivot in star and end
                if(nums[start]>nums[start+1]){
                    return start;
                }
                start++;
                if(nums[end]<nums[end-1]){
                    return end-1;
                }
                end++;

            }
            // left side is sorted so pivot should be in right 
            else if(nums[start]<nums[mid] || (nums[start] == nums[mid] && nums[mid]>nums[end])){
                start = mid+1;
            } else {
                end = mid-1;
            }
             
        }
        return -1;
    }

    int binarySearch(int[] nums, int start, int end, int target){
        while(start <= end){
            int mid = start + (end - start) / 2;
            if(nums[mid] == target){
                return mid;
            } else if(nums[mid] < target){
                start = mid + 1;
            } else{
                end = mid - 1;
            }
        }
        return -1;
    }
}

File: 81. Search in Rotated Sorted Array II/Solution.cpp
class Solution {
public:
    int n;
    int findPivot(vector<int>& nums){
        int l = 0;
        int r = n-1;

        // normal binary search just add few lines to skip duplicates
        while(l<r){
            // skip duplicates from left and right 
            while(l<r && nums[l] == nums[l+1]){
                l++;
            }
            while(r>l && nums[r] == nums[r-1]){
                r--;
            }
            // binary search
            int mid = l + (r-l)/2;
            if(nums[mid]>nums[r]){
                l = mid+1;
            } 
            if(nums[mid]<nums[r]){
                r = mid;
            }
        }
        return r;
    }
    bool binarySearch(vector<int>& nums, int &target,int start, int end){
        while(start<=end){
            int mid = start + (end-start)/2;
            if(nums[mid] == target){
                return true;
            } else if(nums[mid]>target){
                end = mid-1;
            } else {
                start = mid+1;
            }
        }
        return false;
    }
    bool search(vector<int>& nums, int target) {
        n = nums.size();
        int pivot = findPivot(nums);
        if(binarySearch(nums,target,0,pivot-1)){
            return true;
        } else if(binarySearch(nums,target,pivot,n-1)){
            return true;
        } else {
            return false;
        }
    }
};